examples/vulkan/vk_quad_warp/controls.rs

```diff
@@ -0,0 +1,152 @@
+use crate::Model;
+use nannou::ui::prelude::*;
+use nannou::prelude::*;
+use nannou::geom::rect::Rect;
+use self::ui::input::state::mouse::ButtonPosition;
+
+pub const PAD_X: f32 = 20.0;
+pub const PAD_Y: f32 = 20.0;
+
+pub struct Controls {
+    pub corners: Corners,
+}
+
+pub struct Corners {
+    pub dims: Rect<f32>,
+    pub top_left: Corner,
+    pub top_right: Corner,
+    pub bottom_left: Corner,
+    pub bottom_right: Corner,
+}
+
+pub struct Corner {
+    pub drag: bool,
+    pub pos: Point2,
+
+}
+
+pub struct Ids {
+    pub top_left_corner: widget::Id,
+    pub top_right_corner: widget::Id,
+    pub bottom_left_corner: widget::Id,
+    pub bottom_right_corner: widget::Id,
+    pub background: widget::Id,
+    pub points: widget::Id,
+    pub tl_text: widget::Id,
+    pub tr_text: widget::Id,
+    pub bl_text: widget::Id,
+    pub br_text: widget::Id,
+}
+
+impl Corners {
+    pub fn new(init: Rect<f32>) -> Self {
+        Corners {
+            dims: init,
+            top_left: Corner{ drag: false, pos: pt2(init.x.start, init.y.end) },
+            top_right: Corner{ drag: false, pos: pt2(init.x.end, init.y.end) },
+            bottom_left: Corner{ drag: false, pos: pt2(init.x.start, init.y.start) },
+            bottom_right: Corner{ drag: false, pos: pt2(init.x.end, init.y.start) },
+        }
+    }
+}
+
+
+pub(crate) fn update(model: &mut Model) {
+    let ui = &mut model.ui.set_widgets();
+
+    let ref mut corners = model.controls.corners;
+
+    widget::Canvas::new().rgb(0.2, 0.0, 0.2).set(model.ids.background, ui);
+
+    widget::Circle::fill(20.0)
+        .rgb(0.0, 0.7, 0.0)
+        .x_position(position::Position::Absolute(corners.top_left.pos.x as f64))
+        .y_position(position::Position::Absolute(corners.top_left.pos.y as f64))
```

Comment:
you should be able to write these positioning lines as:
```rust
.x(corners.top_left.pos.x as f64)
.y(corners.top_left.pos.y as f64)
```
There are a few widgets below that could also use this.